MDEV-37146 mariadb-dump: Replace SHOW SLAVE STATUS usage with information_schema.SLAVE_STATUS

ParadoxV5 opened the pull request:
* [x] *The Jira issue number for this PR is: [MDEV-37146](https://jira.mariadb.org/browse/MDEV-37146)*

## Description
`SELECT … FROM information_schema.SLAVE_STATUS` is an alternative to `SHOW SLAVE STATUS` that supports `WHERE` and column selection.
Utilizing this server-side feature allows `mariadb-dump --dump-slave` to iterate only the required data on the client side.

The main commit keeps support for pre-`information_schema.SLAVE_STATUS` ([MDEV-33526](https://jira.mariadb.org/browse/MDEV-33526) @ 11.6) via `SHOW SLAVE STATUS` by switching column indices.
On the contrary, it additionally drops `--dump-slave` and `--apply-slave-statements` support for pre-GTID & pre-multi-source (before 10.0 – long EOL).

## Release Notes
`mariadb-dump --dump-slave` now requires MariaDB 10.0 or above, and is additionally optimized for 11.6 or above.

## How can this PR be tested?
Test `mariadb-dump --dump-slave` as usual, e.g., through MTR tests `main.rpl_mysqldump_slave` & `multi_source.mariadb-dump_slave`.

## Basing the PR against the correct MariaDB version
* [x] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
* ~~*This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*~~

## PR quality check
* [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
* [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

ParadoxV5 commented on client/mysqldump.cc:
```diff
@@ -7577,10 +7509,7 @@ int main(int argc, char **argv)
 
   /* Check if the server support multi source */
   if (mysql_get_server_version(mysql) >= 100000)
-  {
-    multi_source= 2;
     have_mariadb_gtid= 1;
-  }
```
### Re: https://github.com/MariaDB/server/pull/3780#discussion_r1923094976

This patch currently drops support for versions that didn’t have multi-source, which are before 10.0.
Is this acceptable because they have long been EOL?

Additionally, to what extent should this program continue to support dumping [MySQL](https://dev.mysql.com/doc/refman/8.0/en/performance-schema-replication-connection-configuration-table.html)?

ParadoxV5 replied:
> Additionally, to what extent should this program continue to support dumping [MySQL](https://dev.mysql.com/doc/refman/8.0/en/performance-schema-replication-connection-configuration-table.html)?

The answer is in [MDEV-39519](https://jira.mariadb.org/browse/MDEV-39519).

ParadoxV5 commented on client/mysqldump.cc:
```diff
@@ -6494,61 +6471,49 @@ static int add_slave_statements(void)
-                                    multi_source ?
-                                    "SHOW ALL SLAVES STATUS" :
-                                    "SHOW SLAVE STATUS"))
+    "SELECT Connection_name, Master_Host, Master_Port,"
+    "Master_Log_File, Exec_Master_Log_Pos FROM information_schema.SLAVE_STATUS")
```
🧐 the index-9 column… This should be
```suggestion
    "SELECT Connection_name, Master_Host, Master_Port, Relay_Master_Log_File,"
    "Exec_Master_Log_Pos FROM information_schema.SLAVE_STATUS")
```

vuvova commented:
Really? I understand that you love `std::function` but this is seriously overdoing it.

bnestere commented on client/mysqldump.cc:
```diff
@@ -174,17 +175,75 @@ static uint opt_use_gtid;
+static MYSQL_RES *routine_res, *routine_list_res= NULL;
+
+static int mysql_query_with_error_report(MYSQL *, MYSQL_RES **, const char *);
+template<typename F> struct Slave_status_list
```
This is a very general interface which applies to pretty much all querying. I'm not necessarily against the idea, but for this use case, I think it doesn't make sense to define here (as once `SSS` support is removed, it would be much less effective). Perhaps it could be discussed with the connector team to add support for something like this.

bnestere reviewed (changes requested):
Overall I agree with @vuvova, I think the overengineering here hurts readability, and I don't really see the benefit of extensbility, as the use case is very niche. Especially when eventually, we will remove support for `SHOW ALL SLAVES STATUS` after 11.4 goes EOL, which makes the new organization somewhat meaningless in the long run. 

Generally, I'd suggest a structure which is tailored for `IS.slave_status`, with easy to delete workarounds w/ TODOs to accommodate old slaves.

Though you did submit a decently sized code re-structuring, so I've left "theoretical" comments on it as things to keep in mind for your future tasks :)

vuvova commented:
I'd do simply
```diff
+  if (mysql_query_with_error_report(mysql_con, &slave_status_res,
+                                    multi_source ?
+                                    "SELECT ... FROM I_S.SLAVE_STATUS ..." :
+                                    "SELECT ... FROM I_S.SLAVE_STATUS ..."))
    if (mysql_query_with_error_report(mysql_con, &slave_status_res,
                                      multi_source ?
                                      "SHOW ALL SLAVES STATUS" :
                                      "SHOW SLAVE STATUS"))
      return(1);
```
and the rest of the code could pretty much stay the same

ParadoxV5 reviewed (commented):
I wasn’t expecting feedback already, @bnestere (and @vuvova), but thank you for the criticism on my approach, nonetheless.

ParadoxV5 committed:
MDEV-37146 mariadb-dump: Replace SHOW SLAVE STATUS usage with information_schema.SLAVE_STATUS

`SELECT … FROM information_schema.SLAVE_STATUS` is an alternative
to `SHOW SLAVE STATUS` that supports `WHERE` and column selection.
Utilizing this server-side feature allows `mariadb-dump --dump-slave`
to iterate only the required data on the client side.

This commit keeps support for pre-`information_schema.SLAVE_STATUS`
(MDEV-33526 @ 11.6) via `SHOW SLAVE STATUS` by switching column indices.
On the contrary, it additionally drops `--dump-slave`
and `--apply-slave-statements` support for pre-GTID &
pre-multi-source (before 10.0 – long EOL).

Reviewed-by: Brandon Nesterenko <brandon.nesterenko@mariadb.com>

ParadoxV5 committed:
Improve `mariadb-dump --dump-slave`’s desc. output

Improve the wording and spacing of the description comment generated
by `mariadb-dump --dump-slave` for its CHANGE MASTER statement.

ParadoxV5 commented on mysql-test/main/rpl_mysqldump_slave.result:
```diff
@@ -204,13 +201,11 @@ CHANGE MASTER TO MASTER_LOG_FILE='slave-bin.000001', MASTER_LOG_POS=BINLOG_START
 --
-
--- A corresponding to the below dump-slave CHANGE-MASTER settings to the slave gtid state is printed later in the file.
+-- The replica GTID position corresponding to the below CHANGE-MASTER settings is printed later in the file.
```
* The previous line calls them *primaries* since before this edit, even though nowhere in the server supports the ‘primary’ term.
  This line might as well match by calling it a *replica*, then.
* This is a position, not a ‘state’.

bnestere commented on mysql-test/main/rpl_mysqldump_slave.result:
```diff
@@ -204,13 +201,11 @@ CHANGE MASTER TO MASTER_LOG_FILE='slave-bin.000001', MASTER_LOG_POS=BINLOG_START
---
--- The following is the SQL position of the replication taken from SHOW SLAVE STATUS at the time of backup.
--- Use this position when creating a clone of, or replacement server, from where the backup was taken.
+-- The following is the replication SQL position taken from SHOW SLAVE STATUS at the time of backup.
```
Sure, keeping `SHOW SLAVE STATUS` in the comment for now is ok by me. I'd suggest filing a JIRA to remove support for SHOW SLAVE STATUS from mariadb-dump after 11.4 goes EOL. _Also note community server doesn't have dependence on enterprise server, but the other way around._

ParadoxV5 replied:
The purpose of supporting `--dump--slave`ing earlier versions is to allow upgrades, which would eventually include ES 11.4 → ES 12.3.

CS 11.4 is not EOL until after half a decade, either.

bnestere reviewed (approved):
Looks good, @ParadoxV5, thank you!

Note I did leave a couple minor suggestions.

ParadoxV5 committed:
MDEV-39190 mariadb-dump --dump-slave outputs partial output for old servers

When dumping from older servers, the MDEV-37146 optimization was
clever to reuse the SHOW ALL SLAVES STATUS from an earlier step.
But it forgot to rewind the tape before playing again,
so it only saw a blank screen.
It also left an extra “SHOW ALL SLAVES STATUS” text in dead code.

This commit fixes both of these oversights.

Reviewed-by: Brandon Nesterenko <brandon.nesterenko@mariadb.com>
